components/custom_board/tas5805m/tas5805m.c

```diff
@@ -0,0 +1,171 @@
+
+esp_err_t tas5805m_ctrl(audio_hal_codec_mode_t mode,
+                        audio_hal_ctrl_t ctrl_state);
+esp_err_t tas5805m_conig_iface(audio_hal_codec_mode_t mode,
+                               audio_hal_codec_i2s_iface_t *iface);
+//static i2c_bus_handle_t i2c_handler;
+
+/*
+ * i2c default configuration
+ */
+/*static i2c_config_t i2c_cfg = {
+    .mode = I2C_MODE_MASTER,
+    .sda_pullup_en = GPIO_PULLUP_ENABLE,
+    .scl_pullup_en = GPIO_PULLUP_ENABLE,
+    .master.clk_speed = 100000,
+};
+*/
+
+/*
+ * Operate fuction of PA
+ */
+audio_hal_func_t AUDIO_CODEC_TAS5805M_DEFAULT_HANDLE = {
+    .audio_codec_initialize = tas5805m_init,
+    .audio_codec_deinitialize = tas5805m_deinit,
+    .audio_codec_ctrl = tas5805m_ctrl,
+    .audio_codec_config_iface = tas5805m_conig_iface,
+    .audio_codec_set_mute = tas5805m_set_mute,
+    .audio_codec_set_volume = tas5805m_set_volume,
+    .audio_codec_get_volume = tas5805m_get_volume,
+    .audio_hal_lock = NULL,
+    .handle = NULL,
+};
+
+esp_err_t tas5805m_read_byte(uint8_t register_name, uint8_t *data)
+{
+
+  int ret;
+  i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | WRITE_BIT, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, register_name, ACK_CHECK_EN);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  if (ret != ESP_OK)
+  {
+    ESP_LOGW(TAG, "I2C ERROR");
+  }
+
+  vTaskDelay(1 / portTICK_RATE_MS);
+  cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | READ_BIT, ACK_CHECK_EN);
+  i2c_master_read_byte(cmd, data, NACK_VAL);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  return ret;
+}
+
+esp_err_t tas5805m_write_byte(uint8_t register_name, uint8_t value)
+{
+  esp_err_t ret;
+  i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | WRITE_BIT, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, register_name, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, value, ACK_CHECK_EN);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  return ret;
+}
+
+esp_err_t
+tas5805m_init(audio_hal_codec_config_t *codec_cfg)
+{
+  // TODO
+  int test = 1;
+  return test;
+}
+
+esp_err_t
+tas5805m_set_volume(int vol)
+{
+  uint8_t volume = vol;
+  return tas5805m_write_byte(TAS5805M_DIG_VOL_CTRL_REGISTER, volume);
+}
+
+esp_err_t tas5805m_get_volume()
+{
+  int ret = tas5805m_read_byte(TAS5805M_DIG_VOL_CTRL_REGISTER, &VolumeLevel);
+  if (ret != ESP_OK)
+  {
+    ESP_LOGW(TAG, "Cant get Volume.");
+    return ret; 
+  }
+  return ret;
+}
+
+esp_err_t
+tas5805m_set_mute()
+{
+  
+  if (*VolumeLevel != TAS5805M_MUTE)
```

Comment:
VolumeLevel is never set to any value or is it?

Reply:
From my understanding it is. If the value != 255 it will be set.

Reply:
Ah, I missed that one